coreneuron/gpu/nrn_acc_manager.cpp

```diff
@@ -43,56 +43,49 @@ void nrn_ion_global_map_delete_from_device();
 void nrn_VecPlay_delete_from_device(NrnThread* nt);
 
-void* cnrn_gpu_copyin(void* h_ptr, std::size_t len) {
+void* cnrn_target_deviceptr(void* h_ptr) {
 #if defined(CORENEURON_ENABLE_GPU) && !defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && defined(_OPENACC)
-    return acc_copyin(h_ptr, len);
+    return acc_deviceptr(h_ptr);
 #elif defined(CORENEURON_ENABLE_GPU) && defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && defined(_OPENMP)
-    auto host_id = omp_get_initial_device();
     auto device_id = omp_get_default_device();
-    auto* d_ptr = omp_target_alloc(len, device_id);
-    nrn_assert(d_ptr != nullptr);
-    nrn_assert(omp_target_memcpy(d_ptr, h_ptr, len, 0, 0, device_id, host_id) == 0);
-    nrn_assert(omp_target_associate_ptr(h_ptr, d_ptr, len, 0, device_id) == 0);
-    return d_ptr;
+    return omp_get_mapped_ptr(h_ptr, device_id);
 #else
-    throw std::runtime_error("cnrn_gpu_copyin() not implemented without OpenACC/OpenMP and gpu build");
+    throw std::runtime_error("cnrn_target_delete() not implemented without OpenACC/OpenMP and gpu build");
 #endif
 }
 
-void cnrn_memcpy_to_device(void* d_ptr, void* h_ptr, size_t len) {
+template <typename T>
+T* cnrn_gpu_copyin(T* h_ptr, std::size_t len = 1) {
 #if defined(CORENEURON_ENABLE_GPU) && !defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && defined(_OPENACC)
-    acc_memcpy_to_device(d_ptr, h_ptr, len);
+    return static_cast<T*>(acc_copyin(h_ptr, len * sizeof(T)));
 #elif defined(CORENEURON_ENABLE_GPU) && defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && defined(_OPENMP)
-    auto host_id = omp_get_initial_device();
-    auto device_id = omp_get_default_device();
-    omp_target_memcpy(d_ptr, h_ptr, len, 0, 0, device_id, host_id);
+    #pragma omp target enter data map(to:h_ptr[:len])
+    return static_cast<T*>(cnrn_target_deviceptr(h_ptr));
 #else
-    throw std::runtime_error("cnrn_memcpy_to_device() not implemented without OpenACC/OpenMP and gpu build");
+    throw std::runtime_error("cnrn_gpu_copyin() not implemented without OpenACC/OpenMP and gpu build");
 #endif
 }
 
-void cnrn_target_delete(void* h_ptr, size_t len) {
+template <typename T>
+void cnrn_target_delete(T* h_ptr, size_t len = 1) {
 #if defined(CORENEURON_ENABLE_GPU) && !defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && defined(_OPENACC)
-    acc_delete(h_ptr, len);
+    acc_delete(h_ptr, len * sizeof(T));
 #elif defined(CORENEURON_ENABLE_GPU) && defined(CORENEURON_PREFER_OPENMP_OFFLOAD) && defined(_OPENMP)
-    (void)len;
-    auto device_id = omp_get_default_device();
-    omp_target_disassociate_ptr(h_ptr, device_id);
-    auto* d_ptr = omp_get_mapped_ptr(h_ptr, device_id);
-    omp_target_free(d_ptr, device_id);
+    #pragma omp target exit data map(delete: h_ptr[:len])
 #else
     throw std::runtime_error("cnrn_target_delete() not implemented without OpenACC/OpenMP and gpu build");
 #endif
 }
 
-void* cnrn_target_deviceptr(void* h_ptr) {
+void cnrn_memcpy_to_device(void* d_ptr, void* h_ptr, size_t len) {
```

Comment:
```suggestion
template <typename T>
void cnrn_memcpy_to_device(T* d_ptr, T* h_ptr, std::size_t len) {
```
would provide a little more type-safety, I think.